pkg/apis/internal.admin.acorn.io/v1/workloadclass.go

```diff
@@ -0,0 +1,113 @@
+package v1
+
+import (
+	"fmt"
+	"math"
+
+	v1 "github.com/acorn-io/acorn/pkg/apis/api.acorn.io/v1"
+	internalv1 "github.com/acorn-io/acorn/pkg/apis/internal.acorn.io/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"github.com/acorn-io/aml"
+)
+
+const gi = 1074000000.0
+
+func ValidateCPU(cfg *v1.Config, app *internalv1.AppInstance, containerName string, container internalv1.Container, spec WorkloadClassSpec, memory resource.Quantity) (resource.Quantity, error) {
+	if err := validateWithMemory(spec, memory, app.Status.Defaults.Memory[containerName]); err != nil {
+		return resource.Quantity{}, err
+	}
+
+	// The CPU scaler calculates the CPUs per Gi of memory so get the memory in a ratio of Gi
+	memoryInGi := memory.AsApproximateFloat64() / (1 * gi)
+	// Since we're putting this in to mili-cpu's, multiply memoryInGi by the scaler and by 1000
+	value := spec.CPUScaler * memoryInGi * 1000
+
+	return *resource.NewMilliQuantity(int64(math.Ceil(value)), resource.DecimalSI), nil
+}
+
+type memoryQuantities struct {
+	Max    *resource.Quantity
+	Min    *resource.Quantity
+	Def    *resource.Quantity
+	Values []*resource.Quantity
+}
+
+func ParseWorkloadClassMemory(memory WorkloadClassMemory) (memoryQuantities, error) {
+	var quantities memoryQuantities
+
+	minInt, err := aml.ParseInt(memory.Min)
+	if err != nil {
+		return memoryQuantities{}, err
+	}
+	quantities.Min = resource.NewQuantity(minInt, resource.BinarySI)
+
+	maxInt, err := aml.ParseInt(memory.Max)
+	if err != nil {
+		return memoryQuantities{}, err
+	}
+	quantities.Max = resource.NewQuantity(maxInt, resource.BinarySI)
+
+	defInt, err := aml.ParseInt(memory.Default)
+	if err != nil {
+		return memoryQuantities{}, err
+	}
+	quantities.Def = resource.NewQuantity(defInt, resource.BinarySI)
+
+	quantities.Values = []*resource.Quantity{}
+	for _, value := range memory.Values {
+		valueInt, err := aml.ParseInt(value)
+		if err != nil {
+			return memoryQuantities{}, err
+		}
+		quantities.Values = append(quantities.Values, resource.NewQuantity(valueInt, resource.BinarySI))
+	}
+
+	return quantities, nil
+}
+
+func validateWithMemory(spec WorkloadClassSpec, memory resource.Quantity, memDefault *int64) error {
+	parsedMemory, err := ParseWorkloadClassMemory(spec.Memory)
+	if err != nil {
+		return err
+	}
+
+	if !memoryInValues(parsedMemory, memory) {
+		// TODO, will want the CC name here
+		return fmt.Errorf("defined memory %v is not an allowed value for the defined WorkloadClass. Allowed values: %v",
+			memory.String(), spec.Memory.Values)
+	}
+
+	memBytes := memory.Value()
+	if parsedMemory.Max.Value() != 0 {
+		switch m := memory.Value(); {
+		case m > parsedMemory.Max.Value():
+			if memBytes == *memDefault {
```

Comment:
It looks like `memBytes` and `m` are the same. Do we need them both?

Reply:
We do not need them both. Good catch!